src/software/util/make_enum/make_enum.hpp

```diff
@@ -0,0 +1,126 @@
+    {                                                                                      \
+    };                                                                                     \
+    template <>                                                                            \
+    constexpr size_t reflective_enum::size<name>()                                         \
+    {                                                                                      \
+        return enum_size_##name;                                                           \
+    }                                                                                      \
+    template <>                                                                            \
+    constexpr auto reflective_enum::values<name>()                                         \
+    {                                                                                      \
+        return enum_values_##name;                                                         \
+    }                                                                                      \
+    template <>                                                                            \
+    constexpr auto reflective_enum::valueNames<name>()                                     \
+    {                                                                                      \
+        return enum_value_names_##name;                                                    \
+    }                                                                                      \
+    inline std::ostream& operator<<(std::ostream& os, name value)                          \
+    {                                                                                      \
+        /* This index lookup relies on the assumption that the enum does not manually */   \
+        /* specify any values. If it did, the underlying integer of the given value */     \
+        /* may be out of range of the vector of strings */                                 \
+        os << reflective_enum::valueNames<name>().at(static_cast<int>(value));             \
+        return os;                                                                         \
+    }
+
+namespace reflective_enum
+{
+    /**
+     * Type trait that checks whether T is a reflective enum, i.e. whether it
+     * was created using the MAKE_ENUM macro.
+     */
+    template <typename T>
+    struct is_reflective_enum : std::false_type
+    {
+    };
+
+    /**
+     * Returns the number of values in the reflective enum E.
+     *
+     * @return the number of values in the enum
+     */
+    template <typename E>
+    constexpr size_t size();
+
+    /**
+     * Returns an array containing the values of the reflective enum E.
+     *
+     * @return an array with the values of the enum
+     */
+    template <typename E>
+    constexpr auto values();
```

Comment:
is `auto` used here because the array size is unknown at compile time?

Reply:
Yes, we don't know the array size right now, but the compiler will figure it out